jsonrpc/server.go

```diff
@@ -192,10 +192,25 @@ func (s *Server) RegisterMethod(method Method) error {
 	return nil
 }
 
-// HandleReader processes a request to the server
+// Handle reads a JSON-RPC request from conn and writes the response.
+func (s *Server) Handle(ctx context.Context, conn io.ReadWriter) error {
+	resp, err := s.handle(ctx, conn)
+	if err != nil {
+		return err
+	}
+	if resp == nil {
+		return nil
+	}
+	if _, err = conn.Write(resp); err != nil {
+		return err
+	}
+	return nil
+}
+
+// handle processes a request to the server
 // It returns the response in a byte array, only returns an
 // error if it can not create the response byte array
-func (s *Server) HandleReader(ctx context.Context, reader io.Reader) ([]byte, error) {
+func (s *Server) handle(ctx context.Context, reader io.Reader) ([]byte, error) {
 	bufferedReader := bufio.NewReader(reader)
```

Comment:
unrelated, but this allocates 4K per request. We should get rid of it.

Reply:
or pool it for reuse

Reply:
Looks like we only use it for determining if the request is a batch. geth does this:

```go
// isBatch returns true when the first non-whitespace characters is '['
func isBatch(raw json.RawMessage) bool {
	for _, c := range raw {
		// skip insignificant whitespace (http://www.ietf.org/rfc/rfc4627.txt)
		if c == 0x20 || c == 0x09 || c == 0x0a || c == 0x0d {
			continue
		}
		return c == '['
	}
	return false
}
```
Could we unmarshal into `json.RawMessage` instead of using the `Decoder`?

Reply:
We can but that will double to allocations, could be worse than what we currently do.

Reply:
Will try to pool it then :+1:

Reply:
Let's just replace the call with `bufio.NewReaderSize(reader, 16)` atleast we allocate less this way.